### PR TITLE
[WIP] S2I tests for demo applications with Quarkus

### PR DESCRIPTION
**Pull request by mjurc**
* Adding a test for TODO Application with Quarkus.

### CODE REVIEW

**Comment by mjurc** [comment]
So, I'm opening this just maybe to get another set of eyes on this cause I am still (slightly) clueless about how to tackle the git s2i with this testsuite.

The issues I am facing currently:

1. I need a dummy app to have a deployment prepared for `app-metadata` extension, which in turn is required for `io.quarkus.ts.openshift:common` extension. I've kinda workarounded it with dummy `helloworld` app from other tests, but I'll probably have to look into the `common` things some more. I kind of have an idea what's required for such S2I build (basically, base image + git repo + context folder), I just am not sure -yet- how to tie it into existing library.
2. The approach I am currently taking is failing on timeouts. As far as I can see, most checks have a [static 5 minute timeout](https://github.com/quarkus-qe/quarkus-openshift-test-suite/blob/master/common/src/main/java/io/quarkus/ts/openshift/common/util/AwaitUtil.java#L51). Maybe... this could be worth parametrising. I am not sure if it would break anything.
3. If the waiters don't time out, the [default wait for ImageStream population](https://github.com/quarkus-qe/quarkus-openshift-test-suite/blob/master/common/src/main/java/io/quarkus/ts/openshift/common/util/AwaitUtil.java#L28) does. This is especially weird considering the [test's additional resources](https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/21/files#diff-b3bca5cbba9c7e7b692951a6e7f49d46R12) already bring that in.

File: todo-demo-app/pom.xml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    <artifactId>todo-app</artifactId>
+    <packaging>jar</packaging>
+
+    <name>Quarkus OpenShift TS: TODO Application with Quarkus</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-container-image-s2i</artifactId>
```
Comment (Ladicek) [inline todo-demo-app/pom.xml:22]: This is brought in automatically via `quarkus-openshift`.

File: todo-demo-app/src/main/resources/application.properties [hunk 2]
```diff
@@ -0,0 +1,5 @@
+quarkus.application.name=todo-demo-dummy-app
+quarkus.http.root-path=/api
+quarkus.kubernetes.deployment-target=openshift
```
Comment (Ladicek) [inline todo-demo-app/src/main/resources/application.properties:3]: You don' need the 3 lines above. If you use `io.quarkus:quarkus-openshift`, then `quarkus.kubernetes.deployment-target` is automatically set to `openshift`. The first 2 lines are just non-default configuration, but defaults should work just fine for you.

File: todo-demo-app/src/test/java/io/quarkus/ts/openshift/tododemoapp/TodoDemoAppIT.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import io.fabric8.openshift.client.OpenShiftClient;
+import io.quarkus.ts.openshift.common.AdditionalResources;
+import io.quarkus.ts.openshift.common.OpenShiftTest;
+import io.quarkus.ts.openshift.common.injection.TestResource;
+import org.junit.jupiter.api.Test;
+
+import static io.restassured.RestAssured.when;
+
+@OpenShiftTest
+@AdditionalResources("classpath:openjdk-11-rhel7.yaml")
```
[the remark this answers is not in the document]
Comment (mjurc) [reply todo-demo-app/src/test/java/io/quarkus/ts/openshift/tododemoapp/TodoDemoAppIT.java:12]: Actually, it isn't:

`openshift.yml`:
```
apiVersion: "image.openshift.io/v1"
kind: "ImageStream"
metadata:
  annotations:
    app.quarkus.io/build-timestamp: "2020-04-24 - 02:26:04 +0000"
    app.opesnhift.io/vcs-url: "https://github.com/mjurc/quarkus-openshift-test-suite.git"
    app.quarkus.io/commit-id: "c147b7b22cd0dc3d351afabf6132c9c771803983"
    app.quarkus.io/vcs-url: "https://github.com/mjurc/quarkus-openshift-test-suite.git"
  labels:
    app.kubernetes.io/name: "todo-demo-dummy-app"
    app.kubernetes.io/version: "1.0.0-SNAPSHOT"
    app.openshift.io/runtime: "quarkus"
  name: "openjdk-11-rhel7"
spec:
  dockerImageRepository: "registry.access.redhat.com/openjdk/openjdk-11-rhel7"
```

`openjdk-11-rhel7.yaml`:
```
apiVersion: image.openshift.io/v1
kind: ImageStream
metadata:
  name: openjdk-11-rhel7
spec:
  lookupPolicy:
    local: false
  tags:
    - name: latest
      from:
        kind: DockerImage
        name: registry.access.redhat.com/openjdk/openjdk-11-rhel7
```

I suppose the `BuildConfig` from yaml could use the one generated by Quarkus, but I'll have to check what's the order in which they are added to OCP.

**Comment by mjurc** [comment]
The application itself builds and functions using `oc apply -f openjdk-11-rhel7.yaml` and `oc apply -f todo-demo-app.yaml`, yes.

I had some ideas too myself, I'm just kinda figuring out how `OpenshiftTestExtension` right now.

**Comment by Ladicek** [comment]
Yea, so first thing probably needs to be something like `@OpenShiftTest(deploymentType = DeploymentType.MANUAL)`, which would be a new thing (default to something like `DeploymentType.BINARY_S2I`, which would be the current behavior). If that is configured, `OpenShiftTestExtension` would skip the application deployment altogether, and instead look for a `public static` method annotated `@ApplicationDeployment`, that would deploy the 2 YAML files you have. That's my current idea, sorry, can't work on it at the moment.

File: todo-demo-app/src/test/java/io/quarkus/ts/openshift/tododemoapp/TodoDemoAppIT.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ladicek) [reply todo-demo-app/src/test/java/io/quarkus/ts/openshift/tododemoapp/TodoDemoAppIT.java:12]: So they are almost the same, and they use the same name, which is certainly not ideal.

File: todo-demo-app/src/test/resources/todo-demo-app.yaml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  metadata:
+    name: todo-demo-app
+  spec:
+    ports:
+    - name: 8080-tcp
+      port: 8080
+      protocol: TCP
+      targetPort: 8080
+    selector:
+      deploymentconfig: todo-demo-app
```
Comment (Ladicek) [inline todo-demo-app/src/test/resources/todo-demo-app.yaml:84]: I wonder how this YAML can work, given that no resource has this `deploymentconfig` label :-)

**Comment by Ladicek** [comment]
@mjurc I came up with something very quickly, see https://github.com/Ladicek/quarkus-openshift-test-suite/commit/86241408c8a8cb388162a865d7790dae13b42d25 and please provide comments :-)

**Comment by mjurc** [comment]
Closing, superseded by https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/25

**Pull request closed without merging**